README
NAME
    Sys::PageCache - handling page cache related on files

INSTALLATION
    To install this module, run the following commands:

        perl Makefile.PL
        make
        make test
        make install

SYNOPSIS
        use Sys::PageCache;
    
        # determine whether pages are resident in memory
        $r = fincore "/path/to/file";
        printf("cached/total_size=%llu/%llu cached/total_pages=%llu/%llu\n",
               $r->{cached_size}, $r->{file_size},
               $r->{cached_pages}, $r->{total_pages},
           );
    
        # free cached pages on a file
        $r = fadvise "/path/to/file", 0, 0, POSIX_FADV_DONTNEED;

DESCRIPTION
    Sys::PageCache is for handling page cache related on files.

METHODS
    fincore($filepath:Str [, $offset:Int [, $length:Int]])
        Determine whether pages are resident in memory. `$offset' and
        `$length' are optional.

        `fincore' returns a following hash ref.

            {
               cached_pages => Int, # number of cached pages
               cached_size  => Int, # size of cached pages
               total_pages  => Int, # number of pages if cached whole file
               file_size    => Int, # size of file
               page_size    => Int, # page size on your system
            }

    fadvise($filepath:Str, $offset:Int, $length:Int, $advice:Int)
        Call posix_fadvise(2).

        `fadvise' returns 1 if success.

    page_size()
        Returns size of page size on your system.

EXPORTS
    fincore
    fadvise
    POSIX_FADV_NORMAL
    POSIX_FADV_SEQUENTIAL
    POSIX_FADV_RANDOM
    POSIX_FADV_NOREUSE
    POSIX_FADV_WILLNEED
    POSIX_FADV_DONTNEED

AUTHOR
    HIROSE Masaaki <hirose31 _at_ gmail.com>

REPOSITORY
    https://github.com/hirose31/Sys-PageCache

      git clone git://github.com/hirose31/Sys-PageCache.git

    patches and collaborators are welcome.

SEE ALSO
    mincore(2), posix_fadvise(2), https://code.google.com/p/linux-ftools/,
    https://github.com/nhayashi/pagecache-tool

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Sys::PageCache

0.02  2013-09-13
    [IMPROVEMENTS]
        - Initial release

0.01_002  2013-09-11
    [IMPROVEMENTS]
        - Support FreeBSD >= 8.3

0.01_001  2013-09-10
    [NEW FEATURES]
        - Initial developers release


t/Util.pm
package t::Util;

use strict;
use warnings;

use base qw(Exporter);
our @EXPORT = qw(p);

use IO::File;
use File::Temp qw(tempfile);
use Data::Dumper;

sub p($) {
    local $Data::Dumper::Indent    = 1;
    local $Data::Dumper::Deepcopy  = 1;
    local $Data::Dumper::Sortkeys  = 1;
    local $Data::Dumper::Terse     = 1;
    local $Data::Dumper::Useqq     = 1;
    local $Data::Dumper::Quotekeys = 0;
    my $d =  Dumper($_[0]);
    $d    =~ s/\\x{([0-9a-z]+)}/chr(hex($1))/ge;
    print STDERR $d;
}

sub create_tempfile {
    my %p = @_;
    my($fh, $filename) = tempfile(UNLINK => 1);
    $fh->autoflush(1);

    if ($p{size} && $p{size} > 0) {
        print {$fh} "X"x$p{size};
        seek $fh, 0, SEEK_SET;
        my $buf = do { local $/; <$fh> };
    }

    return ($fh, $filename);
}

1;

__END__

# for Emacsen
# Local Variables:
# mode: cperl
# cperl-indent-level: 4
# cperl-close-paren-offset: -4
# cperl-indent-parens-as-block: t
# indent-tabs-mode: nil
# coding: utf-8
# End:

# vi: set ts=4 sw=4 sts=0 et ft=perl fenc=utf-8 :


Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;

if ($^O =~ /linux/) {
    ; # OK
} elsif ($^O =~ /freebsd/) {
    my($osver) = (qx(uname -r) =~ /^([0-9.]+)/);
    if (defined $osver && $osver >= 8.3) {
        ; # OK
    } else {
        print "Only support FreeBSD >= 8.3 because of lack of posix_fadvise(2)\n";
        exit 0;
    }
} else {
    print "No support for your OS\n";
    exit 0;
}

my %WriteMakefileArgs = (
    "ABSTRACT_FROM" => "lib/Sys/PageCache.pm",
    "AUTHOR" => "HIROSE Masaaki",
    "BUILD_REQUIRES" => {},
    "CONFIGURE_REQUIRES" => {
        "ExtUtils::MakeMaker" => "6.30"
    },
    "DISTNAME" => "Sys-PageCache",
    "EXE_FILES" => [
        "bin/pagecache-check.pl",
        "bin/pagecache-clear.pl",
    ],
    "LICENSE" => "perl",
    "NAME" => "Sys::PageCache",
    "PREREQ_PM" => {
        "Pod::Usage" => "1.36",
    },
    "TEST_REQUIRES" => {
        "Test::More" => 0
    },
    "VERSION_FROM" => "lib/Sys/PageCache.pm",
    "test" => {
        "TESTS" => "t/*.t"
    },
    LIBS   => [''],
    DEFINE => '',
    INC    => '-I.',
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
    my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
    my $br = $WriteMakefileArgs{BUILD_REQUIRES};
    for my $mod ( keys %$tr ) {
        if ( exists $br->{$mod} ) {
            $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
        } else {
            $br->{$mod} = $tr->{$mod};
        }
    }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
    my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
    my $pp = $WriteMakefileArgs{PREREQ_PM};
    for my $mod ( keys %$br ) {
        if ( exists $pp->{$mod} ) {
            $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
        } else {
            $pp->{$mod} = $br->{$mod};
        }
    }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
    unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Sys/PageCache.pm
package Sys::PageCache;

use strict;
use warnings;
use 5.008001;
use Carp;
use base qw(Exporter);
our @EXPORT = qw(page_size fincore fadvise
                 POSIX_FADV_NORMAL
                 POSIX_FADV_SEQUENTIAL
                 POSIX_FADV_RANDOM
                 POSIX_FADV_NOREUSE
                 POSIX_FADV_WILLNEED
                 POSIX_FADV_DONTNEED
            );
our @EXPORT_OK = qw();

our $VERSION = '0.02';

use POSIX;

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub fincore {
    my($file, $offset, $length) = @_;

    if (! $offset) {
        $offset = 0;
    } elsif ($offset < 0) {
        croak "offset must be >= 0";
    } else {
        my $pa_offset = $offset & ~(page_size() - 1);
        if ($pa_offset != $offset) {
            carp(sprintf "[WARN] offset must be a multiple of the page size so change %llu to %llu",
                 $offset,
                 $pa_offset,
             );
            $offset = $pa_offset;
        }
    }

    my $fsize = (stat $file)[7];
    if (! $length) {
        $length = $fsize;
    } elsif ($length > $fsize - $offset) {
        my $new_length = $fsize - $offset;
        carp(sprintf "[WARN] fincore: length(%llu) is greater than file size(%llu) - offset(%llu). so use file size - offset (=%llu)",
             $length,
             $fsize,
             $offset,
             $new_length,
         );
        $length = $new_length;
    }

    open my $fh, '<', $file or croak $!;
    my $fd = fileno $fh;

    my($r, $e);
    {
        local $@;
        $r = eval {
            _fincore($fd, $offset, $length);
        };
        chomp($e = $@) if $@;
    }
    close $fh;

    if (defined $e) {
        carp $e;
        return;
    }

    $r->{file_size}   = $fsize;
    $r->{total_pages} = ceil($fsize / $r->{page_size});

    return $r;
}

sub fadvise {
    my($file, $offset, $length, $advice) = @_;

    croak "missing advice" unless defined $advice;
    croak "missing length" unless defined $length;
    croak "missing offset" unless defined $offset;
    croak "missing file"   unless defined $file;

    croak "offset must be >= 0" if $offset < 0;

    my $fsize = (stat $file)[7];
    if ($length > $fsize - $offset) {
        my $new_length = $fsize - $offset;
        carp(sprintf "[WARN] fadvise: length(%llu) is greater than file size(%llu) - offset(%llu). so use file size - offset (=%llu)",
             $length,
             $fsize,
             $offset,
             $new_length,
         );
        $length = $new_length;
    }

    open my $fh, '<', $file or croak $!;
    my $fd = fileno $fh;

    my($r, $e);
    {
        local $@;
        $r = eval {
            _fadvise($fd, $offset, $length, $advice);
        };
        chomp($e = $@) if $@;
    }
    close $fh;

    if (defined $e) {
        carp $e;
        return;
    }

    return $r == 0 ? 1 : ();
}

1;
__END__

=encoding utf-8

=begin html

<a href="https://travis-ci.org/hirose31/Sys-PageCache"><img src="https://travis-ci.org/hirose31/Sys-PageCache.png?branch=master" alt="Build Status" /></a>
<a href="https://coveralls.io/r/hirose31/Sys-PageCache?branch=master"><img src="https://coveralls.io/repos/hirose31/Sys-PageCache/badge.png?branch=master" alt="Coverage Status" /></a>

=end html

=head1 NAME

Sys::PageCache - handling page cache related on files

=begin readme

=head1 INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

=end readme

=head1 SYNOPSIS

    use Sys::PageCache;
    
    # determine whether pages are resident in memory
    $r = fincore "/path/to/file";
    printf("cached/total_size=%llu/%llu cached/total_pages=%llu/%llu\n",
           $r->{cached_size}, $r->{file_size},
           $r->{cached_pages}, $r->{total_pages},
       );
    
    # free cached pages on a file
    $r = fadvise "/path/to/file", 0, 0, POSIX_FADV_DONTNEED;

=head1 DESCRIPTION

Sys::PageCache is for handling page cache related on files.

=head1 METHODS

=over 4

=item B<fincore>($filepath:Str [, $offset:Int [, $length:Int]])

Determine whether pages are resident in memory.
C<$offset> and C<$length> are optional.

C<fincore> returns a following hash ref.

    {
       cached_pages => Int, # number of cached pages
       cached_size  => Int, # size of cached pages
       total_pages  => Int, # number of pages if cached whole file
       file_size    => Int, # size of file
       page_size    => Int, # page size on your system
    }


=item B<fadvise>($filepath:Str, $offset:Int, $length:Int, $advice:Int)

Call posix_fadvise(2).

C<fadvise> returns 1 if success.

=item B<page_size>()

Returns size of page size on your system.

=back

=head1 EXPORTS

=over 4

=item fincore

=item fadvise

=item POSIX_FADV_NORMAL

=item POSIX_FADV_SEQUENTIAL

=item POSIX_FADV_RANDOM

=item POSIX_FADV_NOREUSE

=item POSIX_FADV_WILLNEED

=item POSIX_FADV_DONTNEED

=back

=head1 AUTHOR

HIROSE Masaaki E<lt>hirose31 _at_ gmail.comE<gt>

=head1 REPOSITORY

L<https://github.com/hirose31/Sys-PageCache>

  git clone git://github.com/hirose31/Sys-PageCache.git

patches and collaborators are welcome.

=head1 SEE ALSO

mincore(2), posix_fadvise(2),
L<https://code.google.com/p/linux-ftools/>,
L<https://github.com/nhayashi/pagecache-tool>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
